Improve performance find zip archive

johandahlberg opened the pull request:
This is an attempt to address #1662. Working on a code base that makes extensive us of zip archives containing parquet files that need to be read on the fly to load data, I noticed that `find` was a major bottleneck in that process.

I have made a zip specific implementation of `find` here that relies on the file list of the zip file, rather than explicitly walking.

Here is how I measured the performance:

```
from fsspec.implementations.zip import ZipFileSystem

# Example achieve with roughly 900 files, in some deep directory structures
file_system = ZipFileSystem("example.zip")
%timeit file_system.find("/")
```

Performance for find on current `master`
```
2.14 s ± 146 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Performance for find with this fix applied:
```
318 µs ± 11.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

Let me know what you think.

martindurant commented on fsspec/implementations/zip.py:
```diff
@@ -132,3 +132,59 @@ def _open(
+
+        result = {}
+
+        def _below_max_recursion_depth(path):
```
maxdepth really exists for filesystems doing an explicit ls(), to prevent having to walk very deep trees. In this case, we consider all files anyway, so the code can be much simpler by ignoring maxdepth.

johandahlberg replied:
I agree that removing it would make this quite a lot simpler. I wondering, since `maxdepth` is part of the interface if you think that there is any risk that people using it are relying on it only returning files from a specific `maxdepth`? I think that might be a case for keeping it in?

martindurant commented on fsspec/implementations/zip.py:
```diff
@@ -132,3 +132,59 @@ def _open(
+
+        for zip_info in self.zip.infolist():
+            file_name = zip_info.filename
+            if not file_name.startswith(path.lstrip("/")):
```
```suggestion
            if not file_name.startswith(path.lstrip("/")) + "/":
```
? 

So that find('apath') doesn't match 'apathlonger', only 'apath/nested'.

johandahlberg replied:
Yeah. That makes sense.

martindurant commented on fsspec/implementations/zip.py:
```diff
@@ -132,3 +132,59 @@ def _open(
+            # from the file paths
+            if zip_info.is_dir():
+                if withdirs:
+                    if not result.get(file_name) and _below_max_recursion_depth(
```
```suggestion
                    if not file_name in result and _below_max_recursion_depth(
```
seems more obvious

martindurant commented on fsspec/implementations/zip.py:
```diff
@@ -132,3 +132,59 @@ def _open(
+                else:
+                    continue  # Skip along to the next entry if we don't want to add the dirs
+
+            if not result.get(file_name):
```
```suggestion
            if file_name not in result:
```

martindurant commented on fsspec/implementations/zip.py:
```diff
@@ -132,3 +132,59 @@ def _open(
+                        result[file_name.strip("/")] = (
+                            self.info(file_name) if detail else None
+                        )
+                    continue
+                else:
+                    continue  # Skip along to the next entry if we don't want to add the dirs
```
```suggestion
                continue
```
?

martindurant commented on fsspec/implementations/zip.py:
```diff
@@ -132,3 +132,59 @@ def _open(
+
+            if not result.get(file_name):
+                if _below_max_recursion_depth(file_name):
+                    result[file_name] = self.info(file_name) if detail else None
```
How is `info` got here? Don't we know everything about this path at this point?

martindurant commented on fsspec/implementations/zip.py:
```diff
@@ -132,3 +132,59 @@ def _open(
+                if withdirs:
+                    directories = file_name.split("/")
+                    for i in range(1, len(directories)):
+                        dir_path = "/".join(directories[:i]).strip(
+                            "/"
+                        )  # remove the trailing slash, as this is not expected
+                        if not result.get(dir_path) and _below_max_recursion_depth(
+                            dir_path
+                        ):
+                            result[dir_path] = {
+                                "name": dir_path,
+                                "size": 0,
+                                "type": "directory",
+                            }
```
I feel like this code is duplicated from elsewhere. That might be unavoidable, but perhaps worth thinking about refactoring something into utils.

martindurant commented on fsspec/implementations/zip.py:
```diff
@@ -132,3 +132,59 @@ def _open(
+                                "type": "directory",
+                            }
+
+        return result if detail else sorted(result.keys())
```
```suggestion
        return result if detail else sorted(result)
```

johandahlberg commented:
Thank you for the review @martindurant! I refactored the code quite a bit, to make use of the `dir_cache` instead of directly accessing the zip file. I think it made the code easier to follow.

Regarding removing the `maxdepth` that would be a change in the behavior compared to the `find` method on the `AbstractFileSystem` - since my initial strategy here was to try to implement something that had parity in the output I kept it in. If you still want me to remove it let me know and I'll do it (or feel free to push any changes you want to this branch if you prefer that).

I implemented it as suggested by you now, but just filtering for it at the end if it is set.

martindurant commented on fsspec/implementations/zip.py:
```diff
@@ -132,3 +132,53 @@ def _open(
+
+        def _matching_starts(file_path):
+            file_parts = filter(lambda s: bool(s), file_path.split("/"))
+            for a, b in zip(path_parts, file_parts):
```
This could be expressed with `any` or `all` in one line, I think

martindurant commented:
I wonder, should this code go up into AbstractArchiveFileSystem, since it may have a similar performance benefit for libarchive or tar filesystems? I'm not sure if find is slow there or if there's anything else to worry about.

johandahlberg commented on fsspec/implementations/zip.py:
```diff
@@ -132,3 +132,53 @@ def _open(
+                    return False
+            return True
+
+        self._get_dirs()
```
Yeah. It's in the same ball park.

johandahlberg commented:
@martindurant Let me know if you are happy with this or if there is anything else you'd like to see. When it comes to moving to evaluating this libarchive/tar file systems I unfortunately don't think that I will have the bandwidth to take that on.

martindurant commented:
> I unfortunately don't think that I will have the bandwidth to take that on.

Understood - maybe someone else gets the itch if indeed speed is a problem for the other ones.

johandahlberg commented:
@martindurant thank you for a very enjoyable review process, and thanks from brining this in. It will help our use-case immensely.
